modules/mono/glue/cs_files/Quat.cs

```diff
@@ -192,22 +199,56 @@ public Vector3 Xform(Vector3 v)
             return new Vector3(q.x, q.y, q.z);
         }
 
+        // Static Readonly Properties
+        public static Quat Identity { get; } = new Quat(0f, 0f, 0f, 1f);
+
         // Constructors 
         public Quat(real_t x, real_t y, real_t z, real_t w)
         {
             this.x = x;
             this.y = y;
             this.z = z;
             this.w = w;
-        }   
+        }
+
+        public bool IsNormalized()
+        {
+            return Mathf.Abs(LengthSquared - 1) <= Mathf.Epsilon;
+        }
+
         public Quat(Quat q)
-        {                     
-            x = q.x;
-            y = q.y;
-            z = q.z;
-            w = q.w;
+        {
+            this = q;
+        }
+
+        public Quat(Basis basis)
+        {
+            this = basis.Quat();
         }
-        
+
+        public Quat(Vector3 eulerYXZ)
+        {
+            real_t half_a1 = eulerYXZ.y * (real_t)0.5;
+            real_t half_a2 = eulerYXZ.x * (real_t)0.5;
+            real_t half_a3 = eulerYXZ.z * (real_t)0.5;
+
+            // R = Y(a1).X(a2).Z(a3) convention for Euler angles.
+            // Conversion to quaternion as listed in https://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/19770024290.pdf (page A-6)
+            // a3 is the angle of the first rotation, following the notation in this reference.
+
+            real_t cos_a1 = Mathf.Cos(half_a1);
+            real_t sin_a1 = Mathf.Sin(half_a1);
+            real_t cos_a2 = Mathf.Cos(half_a2);
+            real_t sin_a2 = Mathf.Sin(half_a2);
+            real_t cos_a3 = Mathf.Cos(half_a3);
+            real_t sin_a3 = Mathf.Sin(half_a3);
+
+            x = sin_a1 * cos_a2 * sin_a3 + cos_a1 * sin_a2 * cos_a3;
+            y = sin_a1 * cos_a2 * cos_a3 - cos_a1 * sin_a2 * sin_a3;
+            z = -sin_a1 * sin_a2 * cos_a3 + cos_a1 * cos_a2 * sin_a3;
+            w = sin_a1 * sin_a2 * sin_a3 + cos_a1 * cos_a2 * cos_a3;
```

Comment:
For consistency with the C++ counterparts, I think this should be moved to SetEuler, and this constructor should call SetEuler instead of SetEuler calling this constructor.

Reply:
The constructor can't call a method without initialising all field values first.

This means that we essentially have to this if we shift the logic to the method:
```csharp
public Quat(Vector3 eulerYXZ)
{
    this = Quat();
    SetEuler(eulerYXZ);
}
```

Do we want the unnecessary initialisation?

Reply:
You are right, I guess it's fine this way then.